internal/service/advancedclustertpf/validate_schema.go

```diff
@@ -0,0 +1,77 @@
+package advancedclustertpf
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+var defaultMongoDBMajorVersion = "8.0"
+
+func PlanMustUseMongoDBVersion(version float64, mustBeLower bool) FailOnIncompatibleMongoDBVersion {
+	return FailOnIncompatibleMongoDBVersion{
+		Version:     version,
+		MustBeLower: mustBeLower,
+	}
+}
+
+type FailOnIncompatibleMongoDBVersion struct {
+	Version     float64
+	MustBeLower bool
+}
+
+func (v FailOnIncompatibleMongoDBVersion) Description(_ context.Context) string {
+	if v.MustBeLower {
+		return fmt.Sprintf("can only be configured if the mongo_db_major_version is %.1f or lower", v.Version)
+	}
+	return fmt.Sprintf("can only be configured if the mongo_db_major_version is %.1f or higher", v.Version)
+}
+
+func (v FailOnIncompatibleMongoDBVersion) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+func (v FailOnIncompatibleMongoDBVersion) PlanModifyInt64(ctx context.Context, req planmodifier.Int64Request, resp *planmodifier.Int64Response) {
+	if req.ConfigValue.IsUnknown() || req.ConfigValue.IsNull() {
+		return
+	}
+	performValidation(ctx, &req.State, &req.Plan, &resp.Diagnostics, v, req.Path)
+}
+
+func (v FailOnIncompatibleMongoDBVersion) PlanModifyBool(ctx context.Context, req planmodifier.BoolRequest, resp *planmodifier.BoolResponse) {
+	if req.ConfigValue.IsUnknown() || req.ConfigValue.IsNull() {
+		return
+	}
+	performValidation(ctx, &req.State, &req.Plan, &resp.Diagnostics, v, req.Path)
+}
+
+func performValidation(ctx context.Context, state *tfsdk.State, plan *tfsdk.Plan, diags *diag.Diagnostics, v FailOnIncompatibleMongoDBVersion, validationPath path.Path) {
+	var mongoDbMajorVersion types.String
+	var mongoDbMajorVersionState types.String
+	diags.Append(plan.GetAttribute(ctx, path.Root("mongo_db_major_version"), &mongoDbMajorVersion)...)
+	diags.Append(state.GetAttribute(ctx, path.Root("mongo_db_major_version"), &mongoDbMajorVersionState)...)
+	if diags.HasError() {
+		return
+	}
+	mongoDbMajorVersionString := mongoDbMajorVersion.ValueString()
+	if mongoDbMajorVersionString == "" {
+		mongoDbMajorVersionString = mongoDbMajorVersionState.ValueString()
+	}
+	if mongoDbMajorVersionString == "" {
+		mongoDbMajorVersionString = defaultMongoDBMajorVersion
+	}
+	var validationOk bool
+	if v.MustBeLower {
+		validationOk = !IsMajorVersionHigherGreaterOrEqual(&mongoDbMajorVersionString, v.Version)
```

Comment:
Description is inconsistent with validation when MustbeLower is true and versions are equal, is it allowed or not?

Reply:
Good spotted. Fixed in 290b393e5fff8730bcd18f0f501c2d74d3a2469c